src/main/java/org/threadly/litesockets/Client.java

```diff
@@ -3,557 +3,358 @@
-   * This sets the Closer interface for this client.  
-   * Once set the client will call .onClose on it once it a socket close is detected. 
+   * <p>This sets the {@link Closer} interface for this client.  Once set the client will call .onClose 
+   * on it once it a socket close is detected.</p> 
    * 
-   * @param closer sets this clients Closer interface.
+   * @param closer sets this clients {@link Closer} callback.
    */
-  protected void setCloser(Closer closer) {
-    if(! closed.get()) {
-      this.closer = closer;
-    }
-  }
+  public void setCloser(Closer closer);
   
   /**
-   * This is used to notify the clients Reader that a read has been added
-   */
-  protected void callReader() {
-    if(! closed.get() && sei != null && reader != null) {
-      sei.execute(new Runnable() {
-        @Override
-        public void run() {
-          getReader().onRead(Client.this);
-        }});
-    }
-  }
-  
-  /**
-   * Returns the currently set Reader callback.
+   * <p>Returns the currently set Reader callback. </p>
    * 
-   * @return the current Reader for this client.
+   * @return the current {@link Reader} callback for this client.
    */
-  protected Reader getReader() {
-    return reader;
-  }
+  public Reader getReader();
 
   /**
-   * This sets the Reader for the client. 
-   * The Reader is called every time a new read is added to the client
-   * Each time a Reader is called it should call .getRead() *once* on the client
+   * <p>This sets the Reader for the client.This should be set before adding the Client to the {@link SocketExecuterInterface}, 
+   * if its not there is a chance to miss data coming in on the socket.  Once set data received on the socket will be callback 
+   * on this Reader to be processed.</p>
    * 
-   * @param reader the Reader interface to set for this client.
+   * @param reader the {@link Reader} callback to set for this client.
    */
-  protected void setReader(Reader reader) {
-    if(! closed.get()) {
-      this.reader = reader;
-    }
-  }
+  public void setReader(Reader reader);
 
   /**
-   * This allows you to set/change the max buffer size for this client object.
-   * This is the in java memory buffer not the additional socket buffer the OS might setup. 
+   * <p>This allows you to set/change the max buffer size for this client object.
+   * This is the in java memory buffer not the additional socket buffer the OS might setup.</p>
    * 
-   * @param size in bytes.
+   * <p>In general this should be set to the max size you can deal with.  The lower this is the more often we
+   * will end up adding/removing the client from the selectors.  Both the read and write buffers are separate so 
+   * a client can use up to 2x the buffer size set here.  You need room in your heap for buffers for all clients.
+   * These buffers are not kept at full size, so clients will rarely use that much memory if the protocol parsing 
+   * and network are keeping up with the data going in/out.</p>
```

Comment:
It's similar to why we made the PriorityScheduler no longer dynamically scaling.  You may be fine in the healthy case, but you may not be able to handle the worst case.

The idea being that if these buffers lazily grow, you may have accepted more connections than your memory can handle if the buffers grow to their max.  So if you can't service the connections as well (ie CPU load), your buffer growth occurs and you start using more resources per connection.

I don't see many alternatives.  The obvious thing to do would be to allocate the entire buffer from the start, so that the connection is using it's max resources always.  I don't know if I like that or not.

The harder solution would be to do something like the linux kernel does with heap management.  Basically control connection allocations, mark them as needing X heap space, even if they are not using it all.  If you get too many connections for using the heap space (even if your not using it), you deny the connection from establishing.

I like this better, but it only works with the assumption that litesockets is using most (if not all) of the heap space, thus ignoring the rest of the application.  We can take this into account at connection establishment time, but we would not be able to handle the application using more heap later on after the connections were established.  So maybe it's not worth doing at all, since it can't handle all cases.

I mostly wanted to comment that this is something to be aware of.